### ISSUE TITLE
Avoid control contention with multiple Directors

### ISSUE THREAD

**Issue by bilderbuchi** [pymeasure/experiment_control_protocol#14]
In #12, we decided we don't want to restrict an ECP deployment to a single Director instance. 

Use cases are 
* @bmoneke regularly needs more than one director (two or more people working independently on different parts of the beamline).
* Sometimes they need access to the same VISA resource (stepper motor card controlling 6 motors), which is difficult, as only one program may open that resource. The separation of director and actor will allow to use the same card from different directors.

In case we have multiple directors at the same time, we should make sure that they don't "fight" over control of one resource. For example, we could reserve/lock multi-channel Actors to directors on a channel-by-channel basis.

Right now I don't see another usecase than multichannel instruments, as I don't see how to meaningfully "share" control of a single-channel instrument like a spectrometer. A camera or sensor/measurement, though, could be shared (e.g. multiple setups recording room temperature/humidity)?

**Comment by bmoneke** [comment]
> we could reserve/lock multi-channel Actors to directors on a channel-by-channel basis.

Yes, or we leave it open to implement by the user, if someone needs as well locking as multiple access.

> Right now I don't see another usecase than multichannel instruments,

Me neither. For "read access", multiple regular access can be done better via publishing without fear for interference (or multiple access to the device).

**Comment by bklebel** [comment]
In my homebrew implementation of a similar framework, we typically use two Directors at once, distinct from the above mentioned usecases.
One director is solely a GUI, which shows regularly measured data, and can be used to control devices "manually", the other director being a parallel instance for running sequences. The "sequence-director" does its thing typically from the command-line or a script, while the GUI is...well, the GUI, partly for monitoring uses, and partly for manual control. In a different stage of the setup-development, we had the GUI greyed out ("deactivated" in terms of pyQt) while the sequence runs, now with the more distributed framework I did not go for that hassle. 

I think the possibility to lock actors is worthwhile to keep in mind, however the passcode in use should be shareable, to the extent that, e.g. if one notices that the currently running sequence has an issue, one can override the lock, if one knows the passcode. Ideally such interactions would not be necessary, but well, at least in our lab things do not always work in the ideal way ;)

The passcode scheme would prevent unintentional fights, while allowing forced manual control if that may become necessary. A passcode scheme might e.g. deactivate parts of a GUI which are currently controlled by a sequence-director, thus alerting a user to the fact that, if they interfere here, they might break something. In the GUI, one can either actually require a password to be entered (possible in PPMS measuring systems, although I never saw it done), or a simple checkbox to "force manual control", in case there will only ever be two "deliberately fighting" directors, but this would be up to people who implement the protocol. 

I agree with @bmoneke that if it is only about read-access, the best way is to publish all the data multiple other components might need

**Comment by bilderbuchi** [comment]
> One director is solely a GUI, which shows regularly measured data, and can be used to control devices "manually", the other director being a parallel instance for running sequences. The "sequence-director" does its thing typically from the command-line or a script, while the GUI is...well, the GUI, partly for monitoring uses, and partly for manual control.

Thank you for that, I actually did a similar thing, I also had a manual control part and a sequencing part, so you could say I also had two directors. In my case these were both in the same GUI (although the sequencing was running in a separate Thread), so I did not realise this until now.

**Comment by bmoneke** [comment]
I also thought, that there has to be a way to force unlock. Even from someone different than the original locking instance (which might have crashed).

Optional passwords are an option.
However, it might be sufficient to distinguish between unlock (only possible for the one who locked) and force unlock (everyone can do it).

**Comment by bklebel** [comment]
Well, if all the communication comes over the network, then we will have a kind of password once we say "only possible for the one who locked", even if that could be just the name of the respective component - we will have to store that somewhere, and then it does not really matter. 
Still, I think it is a good idea to have the two options to 

1. unlock with name/password
2. force unlock

What we then have to decide is, whether a Component that is unlocked then stays unlocked, or whether we reserve a spot in the header of all messages for the unlocking command, so that it must be sent with every corresponding message in order to go through

**Comment by bmoneke** [comment]
I thought, that a locked node "A" rejects all messages, except those from the node "B", which locked it. If the protocol contains sender names, it is easy to filter for the sender.

"Unlock" makes a node open for all other recipients (to send commands or to lock themselves).
